File: sentil-java/src/main/java/io/github/sedislab/sentil/SafetyFilter.java
package io.github.sedislab.sentil;

/** A least-restrictive shield over a nominal input. */
public final class SafetyFilter extends NativeResource {
    /** A filter enforcing the given bounds, which it consumes. */
    public SafetyFilter(Bounds bounds) throws SentilException {
        super(NativeLib.safetyFilterCreate(bounds.consume()), NativeLib::safetyFilterDestroy);
    }

    /** The input closest to nominal that stays inside the bounds. */
    public double[] filter(double[] nominal) throws SentilException {
        return NativeLib.safetyFilterFilter(handle(), nominal, new double[0], new double[0], 0);
    }

    /**
     * The input closest to nominal that satisfies the bounds and each barrier
     * {@code barrierA[i] . u >= barrierB[i]}.
     */
    public double[] filter(double[] nominal, double[][] barrierA, double[] barrierB)
            throws SentilException {
        int m = barrierA.length;
        int n = nominal.length;
        double[] flat = new double[m * n];
        for (int i = 0; i < m; i++) {
            System.arraycopy(barrierA[i], 0, flat, i * n, n);
        }
        return NativeLib.safetyFilterFilter(handle(), nominal, flat, barrierB, m);
    }
}

File: sentil-matlab/+sentil/SafetyFilter.m
classdef SafetyFilter < handle
    %SAFETYFILTER A least-restrictive shield over a nominal control input.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
    end

    methods
        function obj = SafetyFilter(bounds)
            %SAFETYFILTER A filter over the given bounds, which are consumed.
            obj.Handle = sentil_mex('safety_filter_create', bounds.consume());
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('safety_filter_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
        end

        function u = filter(obj, nominal, barrierA, barrierB)
            %FILTER The input closest to nominal satisfying the bounds and each barrier a_i . u >= b_i.
            obj.assertOpen();
            if nargin < 3
                u = sentil_mex('safety_filter_filter', obj.Handle, double(nominal), [], []);
            else
                u = sentil_mex('safety_filter_filter', obj.Handle, double(nominal), ...
                    double(barrierA).', double(barrierB));
            end
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this filter has been closed');
            end
        end
    end
end